### PR TITLE
[GNA] Limitations refactoring

### PR DESCRIPTION
**Pull request by tadamowicz**
### Details:
- The "Limitations" module is now a singleton.
- The limitations returned depend on the compile target provided during initialization of the "Limitations" object.

### Tickets:
 - 109466

### CODE REVIEW

File: src/plugins/intel_gna/src/backend/gna_limitations.hpp [hunk 1]
```diff
@@ -58,9 +59,41 @@ constexpr uint32_t bytesPerCropElement = 2;
 
 constexpr uint32_t kMemoryPageSize = 4096;
 
+class RuntimeTarget {
```
Comment (marcinkusm) [inline src/plugins/intel_gna/src/backend/gna_limitations.hpp:62]: I think name does not suit do the class purpose.

File: src/plugins/intel_gna/src/backend/gna_limitations.hpp [hunk 2]
```diff
@@ -58,9 +59,41 @@ constexpr uint32_t bytesPerCropElement = 2;
...
+private:
+    RuntimeTarget() {}
+
+    // static std::shared_ptr<Service> m_instance;
+    target::DeviceVersion m_compile_target = target::DeviceVersion::NotSet;
+    size_t m_mem_alignment = 0;
+
+public:
+    static RuntimeTarget& RuntimeTarget::GetInstance() {
```
Comment (marcinkusm) [inline src/plugins/intel_gna/src/backend/gna_limitations.hpp:71]: I see that there are methods in utils, layers and limitiations wich uses this new class in this PR.

This singleton require to be initialized before any method will use it,
I am afraid that it may lead to potential issues in unit tests and in code it self.

Default constructor of the GNAPlugin (used by properties) does not initialize device.

Maybe there is possibility to create instance of this class after device is iniitlized and pass reference or pointer to its instance later where it is needed.

File: src/plugins/intel_gna/src/gna_plugin.cpp [hunk 3]
```diff
@@ -392,6 +392,9 @@ void GNAPlugin::Init() {
 
 void GNAPlugin::InitGNADevice() {
     OV_ITT_SCOPED_TASK(itt::domains::GNA_LT, "InitGNADevice");
+    
+    limitations::RuntimeTarget::GetInstance().Init(config.target->get_effective_compile_target());
```
[the remark this answers is not in the document]
Comment (tadamowicz) [reply src/plugins/intel_gna/src/gna_plugin.cpp]: do_not_review label - It is not supposed to be reviewed

File: src/plugins/intel_gna/src/backend/gna_limitations.hpp (hunk 2 above)
Comment (tadamowicz) [reply src/plugins/intel_gna/src/backend/gna_limitations.hpp]: do_not_review label - It is not supposed to be reviewed

File: src/plugins/intel_gna/src/backend/gna_limitations.hpp (hunk 1 above)
Comment (tadamowicz) [reply src/plugins/intel_gna/src/backend/gna_limitations.hpp]: do_not_review label - It is not supposed to be reviewed

File: src/plugins/intel_gna/src/backend/gna_limitations.hpp [hunk 4]
```diff
@@ -284,178 +153,157 @@ class AbstractValidator {
...
+    constexpr static uint32_t kConvEachKernelByteAlignment = 16;
+    constexpr static uint32_t kInputByteAlignment = 64;
+    constexpr static uint32_t kNoOfInputsDivisor = 8;
+    constexpr static uint32_t kNoOfInputsLowPrecDivisor = 16;
+    constexpr static uint32_t kAffineMaxBatchSize = 8;
+    constexpr static uint32_t kMaxPoolMaxWindowSize = 6;
+    constexpr static uint32_t kCopyMaxGrouping = 8;
+    constexpr static uint32_t kTransposeMaxSize = 65528;
+    // TODO In the future there should be created class/struct representing all limitations for specific device
+    // versions.
```
Comment (sirzabek) [inline src/plugins/intel_gna/src/backend/gna_limitations.hpp:180]: I don't think we need this comment.

File: src/plugins/intel_gna/src/backend/gna_limitations.hpp [hunk 5]
```diff
@@ -284,178 +153,157 @@ class AbstractValidator {
...
-                       const uint32_t kW,
-                       const uint32_t kN,
-                       const uint32_t strideH,
-                       const uint32_t strideW,
-                       const uint32_t dilationH,
-                       const uint32_t dilationW,
-                       OvGnaType inPrecision,
-                       bool exception = true) const override;
+private:
+    Limitations() {}
```
Comment (sirzabek) [inline src/plugins/intel_gna/src/backend/gna_limitations.hpp:285]:
```suggestion
    Limitations() {};
```

File: src/plugins/intel_gna/src/backend/gna_limitations.hpp [hunk 6]
```diff
@@ -284,178 +153,157 @@ class AbstractValidator {
...
+    Limitations(const Limitations&) = delete;
+    Limitations& operator=(const Limitations&) = delete;
+
+    inline const target::DeviceVersion& GetCompileTarget() const {
```
Comment (sirzabek) [inline src/plugins/intel_gna/src/backend/gna_limitations.hpp:289]: Why do we need additional function for getting compile target? Not so long ago the code was refactored to remove all duplicates of target manipulation functions/variables, so this should not be added again here.

File: src/plugins/intel_gna/src/backend/gna_limitations.cpp [hunk 7]
```diff
@@ -655,16 +615,16 @@ bool Validator_35::ShouldUseOnlyConv2DGnaIface() const {
     return true;
 }
 
-std::unique_ptr<AbstractValidator> AbstractValidator::Create(const DeviceVersion& target) {
+std::shared_ptr<AbstractValidator> AbstractValidator::Create(const target::DeviceVersion& target) {
```
Comment (sirzabek) [inline src/plugins/intel_gna/src/backend/gna_limitations.cpp:618]: Why target:: needs to be added here?

File: src/plugins/intel_gna/src/backend/gna_limitations.cpp [hunk 8]
```diff
@@ -485,8 +445,8 @@ const Validator_35::CnnLimits Validator_35::kCnn2DLimits{
     {{1, 255, "kernel height"}, {1, 256, "kernel width"}},                          // kKerneHWlLimit2B
     {{1, 255, "convolution stride height"}, {1, 256, "convolution stride width"}},  // kStrideHWLimit1B
     {{1, 255, "convolution stride height"}, {1, 256, "convolution stride width"}},  // kStrideHWLimit2B
-    {{convDilationHeight, convDilationHeight, "dilation height"},                   // kDilationLimit
-     {convDilationWidth, convDilationWidth, "dilation width"}},
+    {{1, 1, "dilation height"},                                                     // kDilationLimit
+     {1, 1, "dilation width"}},
```
Comment (sirzabek) [inline src/plugins/intel_gna/src/backend/gna_limitations.cpp:449]: Why this change and below?

File: src/plugins/intel_gna/src/backend/gna_limitations.cpp [hunk 9]
```diff
@@ -476,6 +347,95 @@ bool Validator_30::ShouldUseOnlyConv2DGnaIface() const {
     return false;
 }
 
+class Validator_35 : public AbstractValidator {
```
[the remark this answers is not in the document]
Comment (sirzabek) [reply src/plugins/intel_gna/src/backend/gna_limitations.cpp:350]: The question is why we don't have unit tests implemented already and will we ever implement them then?

**Comment by sirzabek** [comment]
General comment is to keep the naming convention - please review the whole PR with this in mind.

File: src/plugins/intel_gna/src/backend/gna_limitations.hpp [hunk 10]
```diff
@@ -284,178 +153,157 @@ class AbstractValidator {
...
+    // Currently split layer only supports 2 bytes in int16 and int8 mode.
+    // In fp32 mode this is not necessary but is useful for testing
+    constexpr static uint32_t kBytesPerSplitElement = 2;
+
+    // Currently crop layer only supports 2 bytes in int16 and int8 mode.
+    // In fp32 mode this is not necessary but is useful for testing
+    constexpr static uint32_t kBytesPerCropElement = 2;
+    constexpr static uint32_t kMemoryPageSize = 4096;
+
+    static inline Limitations& GetInstance() {
```
[the remark this answers is not in the document]
Comment (marcinkusm) [reply src/plugins/intel_gna/src/backend/gna_limitations.hpp:194]: My suggestions:
1. What about creating Factory class which would return pointer to the implementation  of Limitations. It would be up to Factory internal implementation if we return pointer to new instance or to the same instance of the class.
2. It would be even better to split Limitations to group of smaller interfaces split by functionality.

File: src/plugins/intel_gna/tests/unit/gna_memory_alignment.cpp [hunk 11]
```diff
@@ -149,16 +149,14 @@ INSTANTIATE_TEST_SUITE_P(MemoryAlignment_GNA_4_0,
 
 class MemoryAlignmentTest : public ::testing::Test {};
 
-TEST(MemoryAlignmentTest, getMemoryAlignmentBytes_ExpectExceptionWhenTargetIsUnset) {
-    EXPECT_ANY_THROW(getMemoryAlignmentBytes(DeviceVersion::NotSet));
-}
-
 TEST(MemoryAlignmentTest, getMemoryAlignmentBytes_Expect64ByteAlignmentWhenTargetIsGNA3_0) {
-    EXPECT_EQ(getMemoryAlignmentBytes(DeviceVersion::GNA3_0), 64);
+    Limitations::GetInstance().Init(DeviceVersion::GNA3_5);
```
Comment (rjeziers) [inline src/plugins/intel_gna/tests/unit/gna_memory_alignment.cpp:153]: If you intended to change GNA3_0 to GNA3_5, please also update the test case name

**Review by rjeziers — CHANGES_REQUESTED** [review]

File: src/plugins/intel_gna/tests/unit/gna_memory_alignment.cpp [hunk 12]
```diff
@@ -149,16 +149,14 @@ INSTANTIATE_TEST_SUITE_P(MemoryAlignment_GNA_4_0,
 
 class MemoryAlignmentTest : public ::testing::Test {};
 
-TEST(MemoryAlignmentTest, getMemoryAlignmentBytes_ExpectExceptionWhenTargetIsUnset) {
-    EXPECT_ANY_THROW(getMemoryAlignmentBytes(DeviceVersion::NotSet));
```
[the remark this answers is not in the document]
Comment (tadamowicz) [reply src/plugins/intel_gna/tests/unit/gna_memory_alignment.cpp:153]: fixed

File: src/plugins/intel_gna/src/backend/gna_limitations.hpp (hunk 5 above)
Comment (tadamowicz) [reply src/plugins/intel_gna/src/backend/gna_limitations.hpp]: General rule is not to add a semicolon ';' at the end of method implementations within a class. BTW I will move implementation to cpp.

File: src/plugins/intel_gna/src/backend/gna_limitations.hpp (hunk 6 above)
Comment (tadamowicz) [reply src/plugins/intel_gna/src/backend/gna_limitations.hpp]: This is a private method of the Limitations class used in the Init function for validating the compile target. The reason m_compile_target cannot be used directly is that Init is a static type.

File: src/plugins/intel_gna/tests/unit/gna_memory_alignment.cpp (hunk 11 above)
Comment (tadamowicz) [reply src/plugins/intel_gna/tests/unit/gna_memory_alignment.cpp]: ok

File: src/plugins/intel_gna/src/gna_plugin.cpp [hunk 13]
```diff
@@ -368,13 +370,19 @@ GNAPlugin::GNAPlugin() : graphCompiler(config) {
     Init();
     UpdateFieldsFromConfig();
     InitGNADevice();
+    Limitations::GetInstance().Init(this->config.target->get_effective_compile_target());
```
[the remark this answers is not in the document]
Comment (tadamowicz) [reply src/plugins/intel_gna/src/gna_plugin.cpp]: ok

File: src/plugins/intel_gna/src/backend/gna_limitations.cpp (hunk 7 above)
Comment (tadamowicz) [reply src/plugins/intel_gna/src/backend/gna_limitations.cpp:618]: will be removed

File: src/plugins/intel_gna/src/backend/gna_limitations.hpp (hunk 4 above)
Comment (tadamowicz) [reply src/plugins/intel_gna/src/backend/gna_limitations.hpp:181]: ok

File: src/plugins/intel_gna/tests/unit/transformations/gna_decompose_2d_convolution.cpp [hunk 14]
```diff
@@ -779,7 +784,7 @@ static size_t CalculateConvCount(const ConvParams& conv_params) {
     size_t conv_count = 1;
     size_t total_factorized_conv_channel_count =
         (conv_params.input_channel_count * conv_params.filter_height * conv_params.filter_width);
-    while (total_factorized_conv_channel_count / conv_count > ov::intel_gna::limitations::convFilterMaxSize ||
+    while (total_factorized_conv_channel_count / conv_count > Limitations::kConvFilterMaxSize ||
```
[the remark this answers is not in the document]
Comment (tadamowicz) [reply src/plugins/intel_gna/tests/unit/transformations/gna_decompose_2d_convolution.cpp:796]: let's do this in separate PR.

File: src/plugins/intel_gna/tests/unit/gna_memory_alignment.cpp [hunk 15]
```diff
@@ -45,8 +45,8 @@ class GNAPluginForMemoryAlignmentTest : public GNAPlugin {
     GNAPluginForMemoryAlignmentTest(const std::map<std::string, std::string>& configMap) : GNAPlugin(configMap) {
         if (gnadevice) {
             gnamem.reset(new gna_memory_float(memory::GNAFloatAllocator{},
-                                              gnadevice->getMemAlignment(),
-                                              limitations::kMemoryPageSize));
+                                              Limitations::GetInstance().GetMemoryAlignment(),
+                                              Limitations::kMemoryPageSize));
```
[the remark this answers is not in the document]
Comment (tadamowicz) [reply src/plugins/intel_gna/tests/unit/gna_memory_alignment.cpp:49]: let's do this in separate PR.

File: src/plugins/intel_gna/tests/unit/gna_get_aligned_split_sizes.cpp [hunk 16]
```diff
@@ -59,7 +59,7 @@ TEST(CheckSplitSupported, CheckVariadicSplitSupported) {
             ngraph::opset9::Constant::create(ngraph::element::i64,
                                              ngraph::Shape({split_lengths.size()}),
                                              split_lengths));
-        ASSERT_TRUE(ov::intel_gna::limitations::is_split_supported(split, false) == result);
+        ASSERT_TRUE(ov::intel_gna::limitations::Limitations::Limitations::IsSplitSupported(split, false) == result);
```
[the remark this answers is not in the document]
Comment (tadamowicz) [reply src/plugins/intel_gna/tests/unit/gna_get_aligned_split_sizes.cpp:62]: fixed

File: src/plugins/intel_gna/src/backend/gna_limitations.hpp [hunk 17]
```diff
@@ -284,178 +153,157 @@ class AbstractValidator {
...
+     * @brief Validates if legacy convolution is supported by GNA
+     * @param conv_ie convolution
+     * @param gna_precision GNA inference precision
+     * @param is_exception_allowed flag specifies whether exception is allowed
+     * @return true if supported
+     */
+    bool IsConvSupported(const std::shared_ptr<ngraph::op::ConvolutionIE>& conv_ie,
+                         const InferenceEngine::Precision gna_precision,
+                         bool is_exception_allowed = false);
+    /**
```
[the remark this answers is not in the document]
Comment (tadamowicz) [reply src/plugins/intel_gna/src/backend/gna_limitations.hpp:247]: done

File: src/plugins/intel_gna/src/backend/gna_limitations.hpp [hunk 18]
```diff
@@ -284,178 +153,157 @@ class AbstractValidator {
...
+     * @brief Check if all operations are supported by GNA
+     * @param model ngraph model
+     * @param gna_compile_target GNA compile target
+     * @param gna_precision GNA inference precision
+     */
+    void CheckAllOpsSupported(const std::shared_ptr<ov::Model>& model, const InferenceEngine::Precision gna_precision);
+    bool UseOnly16BitConvolutionWeights() const;
+
+    inline bool IsCropAffinedOffset(size_t numberOfElements) const {
```
[the remark this answers is not in the document]
Comment (tadamowicz) [reply src/plugins/intel_gna/src/backend/gna_limitations.hpp:277]: For inline methods I am providing definition in header file to avoid linking problems - unresolved external symbols

File: src/plugins/intel_gna/src/backend/gna_limitations.hpp [hunk 19]
```diff
@@ -284,178 +153,157 @@ class AbstractValidator {
...
-                       const uint32_t kW,
-                       const uint32_t kN,
-                       const uint32_t strideH,
-                       const uint32_t strideW,
-                       const uint32_t dilationH,
-                       const uint32_t dilationW,
-                       OvGnaType inPrecision,
-                       bool exception = true) const override;
-};
+    constexpr static uint32_t kBufferMaxSize = 65528;
```
[the remark this answers is not in the document]
Comment (tadamowicz) [reply src/plugins/intel_gna/src/backend/gna_limitations.hpp]: moved to the end of public section

File: src/plugins/intel_gna/src/backend/gna_limitations.hpp [hunk 20]
```diff
@@ -284,178 +153,157 @@ class AbstractValidator {
...
+    constexpr static uint32_t kMemoryPageSize = 4096;
+
+    static inline Limitations& GetInstance() {
+        static Limitations instance;
+        return instance;
+    }
+
+    void Init(const target::DeviceVersion& compile_target);
+
+    static bool IsTranspose2d(const std::vector<size_t>& shape);
```
[the remark this answers is not in the document]
Comment (tadamowicz) [reply src/plugins/intel_gna/src/backend/gna_limitations.hpp]: done

File: src/plugins/intel_gna/src/backend/gna_limitations.cpp (hunk 7 above)
[the remark this answers is not in the document]
Comment (tadamowicz) [reply src/plugins/intel_gna/src/backend/gna_limitations.cpp]: Currently, the validatar is used by GraphCompiler and one of the transformations. We have agreed offline to leave it as it is.

File: src/plugins/intel_gna/src/backend/gna_limitations.hpp (hunk 10 above)
[the remark this answers is not in the document]
Comment (tadamowicz) [reply src/plugins/intel_gna/src/backend/gna_limitations.hpp]: add 1. We agreed that a singleton implementation is a sufficient solution, consequently allowing only one instance of the plugin to exist at the same time.
add 2. I agree that it would be better, but I would like to limit the number of changes for a single pull request. However, this idea can be considered for future improvements.

File: src/plugins/intel_gna/src/backend/gna_limitations.cpp [hunk 21]
```diff
@@ -37,250 +37,62 @@ namespace intel_gna {
...
-        }
-    }
-    if (!error.str().empty()) {
-        THROW_GNA_EXCEPTION << error.str();
-    }
-}
 namespace cnn2d {
 
-bool IsEqualToLimit::isValid(const uint32_t val) const {
+bool IsEqualToLimit::IsValid(const uint32_t val) const {
```
[the remark this answers is not in the document]
Comment (tadamowicz) [reply src/plugins/intel_gna/src/backend/gna_limitations.cpp:56]: This is not covered by the scope of refactoring.

File: src/plugins/intel_gna/src/backend/gna_limitations.cpp (hunk 9 above)
[the remark this answers is not in the document]
Comment (tadamowicz) [reply src/plugins/intel_gna/src/backend/gna_limitations.cpp:350]: This is not covered by the scope of refactoring.

File: src/plugins/intel_gna/src/gna_graph_compiler.cpp [hunk 22]
```diff
@@ -1527,7 +1525,7 @@ void GNAGraphCompiler::GemmPrimitive(InferenceEngine::CNNLayerPtr layer) {
     const auto out_dims = outputs->getDims();
     const auto out_dims_size = ngraph::shape_size(out_dims);
     uint32_t num_rows_out = InferenceEngine::GetDimFromBack(out_dims, 1);
-    uint32_t num_padding = ALIGN(num_rows_in, noOfInputsDivisor) - num_rows_in;
+    uint32_t num_padding = ALIGN(num_rows_in, kNoOfInputsDivisor) - num_rows_in;
```
Comment (marcinkusm) [inline src/plugins/intel_gna/src/gna_graph_compiler.cpp:1528]: it is inconsistent - please use one convention for whole PR or at least for the same file.
In one case There is used `Limitations::` in other not.

image

image

File: src/plugins/intel_gna/src/gna_plugin.cpp [hunk 23]
```diff
@@ -376,27 +384,36 @@ void GNAPlugin::Init() {
...
+                                                     Limitations::kMemoryPageSize);
     }
-    graphCompiler.setGNAMemoryPtr(gnamem);
+}
+
+void GNAPlugin::InitGraphCompiler() {
+    OV_ITT_SCOPED_TASK(itt::domains::GNA_LT, "InitGraphCompiler");
+
+    m_graph_compiler = std::make_shared<GNAGraphCompiler>(
+        GNAGraphCompiler(config, dnn, inputs_ptr_, Limitations::get_instance()->get_cnn_validator(), gnamem));
```
Comment (marcinkusm) [inline src/plugins/intel_gna/src/gna_plugin.cpp:416]: is formatting applied properly?

File: src/plugins/intel_gna/src/transformations/convert_matmul_to_pointwise_convolution.cpp [hunk 24]
```diff
@@ -49,9 +50,9 @@ static std::tuple<bool, uint32_t, uint32_t, uint32_t> VerifyAndGetConvParams(
     const uint32_t width = input1_shape.front();
     const uint32_t in_channels = input2_shape.back();
     const uint32_t out_channels = input2_shape.front();
-    if (input1_shape.front() <= limitations::affineMaxBatchSize ||
-        out_channels % limitations::convFiltersNumDivider != 0 || out_channels > limitations::convMaxFiltersNum ||
-        in_channels > limitations::convFilterMaxSize) {
+    if (input1_shape.front() <= Limitations::kAffineMaxBatchSize ||
```
Comment (marcinkusm) [inline src/plugins/intel_gna/src/transformations/convert_matmul_to_pointwise_convolution.cpp:53]: if it is used in more than one place please create method for this.

File: src/plugins/intel_gna/tests/deprecated/unit/engines/gna/I8_quantisation_test.cpp [hunk 25]
```diff
@@ -9,8 +9,10 @@
 #include "frontend/layer_quantizer.hpp"
 #include "gna_matcher.hpp"
 #include <ie_core.hpp>
+#include <backend/gna_limitations.hpp>
```
Comment (marcinkusm) [inline src/plugins/intel_gna/tests/deprecated/unit/engines/gna/I8_quantisation_test.cpp:12]: please use local scope includes for openvino code.

File: src/plugins/intel_gna/tests/deprecated/unit/engines/gna/i16_quantisation_test.cpp [hunk 26]
```diff
@@ -11,8 +11,10 @@
 #include "gna_matcher.hpp"
 #include <ie_core.hpp>
 #include "ngraph_functions/builders.hpp"
+#include <backend/gna_limitations.hpp>
```
Comment (marcinkusm) [inline src/plugins/intel_gna/tests/deprecated/unit/engines/gna/i16_quantisation_test.cpp:14]: please use local scope includes for openvino code. The same for other files.

**Review by marcinkusm — COMMENTED** [review]
Great Work.
looks fine.
I have added few comments before I will be able to approve.

File: src/plugins/intel_gna/src/backend/gna_limitations.cpp [hunk 27]
```diff
@@ -705,15 +650,293 @@ bool AbstractValidator::ValidationSuccesful(const bool throwOnError,
...
+    if (k_instance) {
+        if (get_instance()->get_compile_target() != compile_target) {
+            THROW_GNA_EXCEPTION << "Limitations instance already initialized with a different configuration.\n";
+        }
+    } else {
+        k_instance = std::shared_ptr<Limitations>(new Limitations(compile_target));
+    }
+}
+
+void Limitations::reset() {
```
[the remark this answers is not in the document]
Comment (tadamowicz) [reply src/plugins/intel_gna/src/backend/gna_limitations.cpp:696]: ok

File: src/plugins/intel_gna/src/backend/am_intel_dnn.cpp [hunk 28]
```diff
@@ -180,8 +182,8 @@ void AMIntelDNN::InitConvolutional1DComponentPrivate(intel_dnn_component_t& comp
         THROW_GNA_EXCEPTION << "Number of inputs to Convolutional1DComponent (" << num_columns_in
                             << ") is not a multiply by 8";
     }
-    if (num_filters < limitations::convMinFiltersNum || num_filters > limitations::convMaxFiltersNum ||
-        num_filters % limitations::convFiltersNumDivider != 0) {
+    if (num_filters < Limitations::kConvMinFiltersNum || num_filters > Limitations::kConvMaxFiltersNum ||
```
[the remark this answers is not in the document]
Comment (tadamowicz) [reply src/plugins/intel_gna/src/backend/am_intel_dnn.cpp:185]: out of PR's scope

File: src/plugins/intel_gna/src/gna_graph_compiler.cpp (hunk 22 above)
Comment (tadamowicz) [reply src/plugins/intel_gna/src/gna_graph_compiler.cpp:1528]: ok

File: src/plugins/intel_gna/src/gna_plugin.cpp (hunk 23 above)
Comment (tadamowicz) [reply src/plugins/intel_gna/src/gna_plugin.cpp:416]: I am using configuration from our coding style guide, also CI does not report any issues

File: src/plugins/intel_gna/src/backend/gna_limitations.hpp (hunk 18 above)
[the remark this answers is not in the document]
Comment (marcinkusm) [reply src/plugins/intel_gna/src/backend/gna_limitations.hpp]: but if you add inline in source file it should not be problem.

in header 
```
bool IsCropAffinedOffset(size_t numberOfElements) const;
```
in source file
```
inline bool Limitations::IsCropAffinedOffset(size_t numberOfElements) const {
//
}
```

File: src/plugins/intel_gna/src/transformations/convert_matmul_to_pointwise_convolution.cpp (hunk 24 above)
Comment (tadamowicz) [reply src/plugins/intel_gna/src/transformations/convert_matmul_to_pointwise_convolution.cpp:53]: out of PR scope

**Comment by tadamowicz** [comment]
> General comment is to keep the naming convention - please review the whole PR with this in mind.

OK

File: src/plugins/intel_gna/tests/deprecated/unit/engines/gna/I8_quantisation_test.cpp (hunk 25 above)
Comment (tadamowicz) [reply src/plugins/intel_gna/tests/deprecated/unit/engines/gna/I8_quantisation_test.cpp:12]: ok

File: src/plugins/intel_gna/src/backend/gna_limitations.hpp (hunk 6 above)
Comment (tadamowicz) [reply src/plugins/intel_gna/src/backend/gna_limitations.hpp]: as confirmed offline removig compile target Limitations fields

File: src/plugins/intel_gna/src/backend/gna_limitations.hpp (hunk 18 above)
[the remark this answers is not in the document]
Comment (tadamowicz) [reply src/plugins/intel_gna/src/backend/gna_limitations.hpp]: done

File: src/plugins/intel_gna/src/backend/gna_limitations.cpp [hunk 29]
```diff
@@ -395,6 +190,66 @@ std::string RectLimitByChannelsAndPrecision::GetErrorOrEmpty(const uint32_t h,
     return GetByPrecision(precision).GetErrorOrEmpty(h, w, channels, what);
 }
 
+class Validator_30 : public AbstractValidator {
+    static const RangeLimit2D kInputHWLimit;
```
Comment (marcinkusm) [inline src/plugins/intel_gna/src/backend/gna_limitations.cpp:194]: I would put private at the end of declaration order.

File: src/plugins/intel_gna/src/backend/gna_limitations.cpp [hunk 30]
```diff
@@ -493,6 +349,95 @@ bool Validator_30::ShouldUseOnlyConv2DGnaIface() const {
...
+        const RangeLimit2D kKerneHWlLimit1B;
+        const RangeLimit2D kKerneHWlLimit2B;
+        const RangeLimit2D kStrideHWLimit1B;
+        const RangeLimit2D kStrideHWLimit2B;
+        const RangeLimit2D kDilationLimit;
+        const RangeLimit2D kPoolingWindowHWLimit;
+        const RangeLimit2D kPoolingStrideHWLimit;
+    };
+
+    static const CnnLimits kCnn2DLimits;
```
Comment (marcinkusm) [inline src/plugins/intel_gna/src/backend/gna_limitations.cpp:367]: I would put privarte fields and methods at the end of declaration order

**Review by marcinkusm — APPROVED** [review]

File: src/plugins/intel_gna/src/backend/gna_limitations.cpp (hunk 29 above)
Comment (tadamowicz) [reply src/plugins/intel_gna/src/backend/gna_limitations.cpp:194]: AbstractValidator class is out of refactoring scope for this PR, I focused only on Limitations class

File: src/plugins/intel_gna/src/backend/gna_limitations.cpp (hunk 30 above)
Comment (tadamowicz) [reply src/plugins/intel_gna/src/backend/gna_limitations.cpp:367]: The same as above, I applied our coding standard rules only for Limitations class

File: src/plugins/intel_gna/src/gna_plugin.cpp [hunk 31]
```diff
@@ -1700,7 +1712,8 @@ InferenceEngine::QueryNetworkResult GNAPlugin::QueryNetwork(
                 TransformationsPipeline(qn_config).apply(model);
             },
             [&](const std::shared_ptr<ngraph::Node>& op) {
-                return limitations::is_op_supported(op, effectiveCompileTarget, qn_config.gnaPrecision);
+                const auto res = Limitations::get_instance()->is_op_supported(op, qn_config.gnaPrecision);
+                return res;
```
Comment (sirzabek) [inline src/plugins/intel_gna/src/gna_plugin.cpp:1716]: Why do we need res?

**Review by sirzabek — APPROVED** [review]

**Review by rjeziers — APPROVED** [review]
